File: silver/longest_good_segment.py
# https://www.codechef.com/problems/LGSEG

#%%
def brute_force(nums,N,K,S):
    def can_split(arr):
        curr,seg = arr[0],1
        for num in arr[1:]:
            if curr + num > S:
                seg += 1
                curr = 0
            curr += num
        return seg <= K
    res = -float('inf')
    for i in range(N):
        for j in range(i+1,N+1):
            if can_split(nums[i:j]):
                res = max(res,j-i)
    return res


def solve1(nums,N,K,S):
    def compute(i):
        seg,curr,j = 1,nums[i],i+1
        while seg <= K and j<N:
            if curr + nums[j] > S:
                curr = 0
                seg += 1
            curr += nums[j]
            j += 1
        return j-i-(seg>K)
    return max(compute(i) for i in range(N))

def test(solve,brute_force = brute_force):
    from random import randint
    for _ in range(50):
        nums = [randint(10,100) for _ in range(100)]
        K = randint(1,10)
        S = max(nums)
        if solve(nums,len(nums),K,S) != brute_force(nums,len(nums),K,S):
            print(nums,K,S)
            print('expected: ', brute_force(nums,len(nums),K,S))
            print('actual: ', solve(nums,len(nums),K,S))
            break
    else:
        print("SUCCESS")
# %%
def build(N,start_index):
    up = [[None]*N for _ in range(20)]
    up[0] = start_index
    for i in range(1,20):
        for j in range(N):
            p = up[i-1][j]
            if p == -1:
                up[i][j] = -1
            else:
                up[i][j] = up[i-1][p]
    return up

def call(up, node, K):
    last,jump = node,1
    for i in range(19):
        if node == -1: break
        if K & jump:
            node = up[i][node]
        jump <<= 1
    return last-node

def solve2(nums,N,K,S):
    start_index,j,curr = [],0,0
    for i in range(N):
        curr += nums[i]
        while curr > S:
            curr -= nums[j]
            j += 1
        start_index.append(j-1)
    
    up = build(N,start_index)
    res = 0
    for i in range(N-1,-1,-1):
        res = max(res, call(up,i,K))
    return res
# %%

File: silver/longest_good_segment.cpp
#include <bits/stdc++.h>
using namespace std;


#define getT int t;cin>>t;while(t--)
#define ipair pair<int,int>

vector<int> arr;
vector<int> startIndex;
vector<int> table[20];

void build(int n)
{
    table[0]=startIndex;
    for(int i=1;i<20;i++)
    {
        table[i]=vector<int>(n);
        for(int j=0;j<n;j++)
        {
            int p=table[i-1][j];
            if(p==-1)
                table[i][j]=-1;
            else
                table[i][j]=table[i-1][p];
        }
    }
}

int call(int node,int k)
{
    int last=node;
    int jump=1;
    
    for(int i=0;i<19 && node>-1;i++)
    {
        if(jump&k)
        {
            node=table[i][node];
        }
        jump<<=1;
    }
    return last-node;
    
}

int main() {
    ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	
	getT{
	    int n,k,s;
	    cin>>n>>k>>s;
	    
	    arr=vector<int>(n);
	    startIndex=vector<int>(n);
	    for(int i=0;i<n;i++)
	        cin>>arr[i];
	   
	    long sum=0;
	    for(int i=0,j=0;i<n;i++)
	    {
	        sum+=arr[i];
	        while(sum>s)
	        {
	            sum-=arr[j];
	            j++;
	        }
	        startIndex[i]=j-1;
	    }
	    
	    build(n);
	    
	    int res=0;
	    for(int i=n-1;i>=0;i--)
	        res=max(res,call(i,k));
	    
	    cout<<res<<"\n";
	    
	}
	
	
	return 0;
}
